src/main/java/edu/umich/its/cpm/MigrationTaskService.java

```diff
@@ -1121,5 +1125,318 @@ private String getExistingBoxFolderIdFromBoxException(BoxAPIException e,
+					String ioError = "downloadMailArchiveZipFile: problem adding zip folder for MailArchive channel "
+							+ channelName + " " + e;
+					log.error(ioError);
+				}
+			}
+			
+			// get all email messages in the channel
+			requestUrl = env.getProperty(Utils.ENV_PROPERTY_CTOOLS_SERVER_URL)
+					+ "direct/mailarchive/channelMessages/" + site_id + "/" + channelId + ".json?_sessionId="
+					+ sessionId;
+			JSONObject messagesJSON = new JSONObject(restTemplate.getForObject(requestUrl,
+					String.class));
+			JSONArray messages = messagesJSON.getJSONArray(Utils.JSON_ATTR_MAILARCHIVE_COLLECTION);
+			for (int iMessage = 0; iMessage < messages.length(); iMessage++) {
+				JSONObject message = messages.getJSONObject(iMessage);
+				
+				// create file for each message
+				String messageFolderName = getMailArchiveMessageFolderName(message, channelName, folderForChannels);
+						
+				JSONObject messageStatus = new JSONObject();
+				messageStatus.put(Utils.JSON_ATTR_MAIL_MESSAGE, messageFolderName);
+				
+				// 1. write the message file
+				messageStatus = handleMailArchiveMessage(out, message, messageFolderName,
+						messageStatus);
+				
+				// 2. get attachments, if any
+				messageStatus = handleMailArchiveMessageAttachments(
+						sessionId, httpContext, out, message, messageFolderName, messageStatus);
+				
+				messagesStatus.put(messageStatus);
+			}
+		}
+		
+		downloadStatus.put(Utils.MIGRATION_DATA, messagesStatus);
+		downloadStatus.put(Utils.MIGRATION_STATUS, messagesStatus.toString().contains(Utils.STATUS_FAILURE)? Utils.STATUS_FAILURE: Utils.STATUS_SUCCESS);
+
+		return downloadStatus;
+	}
+
+	/**
+	 * output MailArchive message content
+	 * @param out
+	 * @param message
+	 * @param messageFolderName
+	 * @param messageStatus
+	 * @return
+	 * @throws IOException
+	 */
+	private JSONObject handleMailArchiveMessage(ZipOutputStream out,
+			JSONObject message, String messageFolderName,
+			JSONObject messageStatus) throws IOException {
+		try {
+			// get the html file content first
+			String messageContent = message.has(Utils.JSON_ATTR_MAIL_BODY)?message.getString(Utils.JSON_ATTR_MAIL_BODY):"";
+
+			ZipEntry fileEntry = new ZipEntry(messageFolderName + Utils.MAIL_MESSAGE_FILE_NAME);
+			out.putNextEntry(fileEntry);
+			
+			// output message header info
+			JSONArray headers = message.getJSONArray(Utils.JSON_ATTR_MAIL_HEADERS);
+			for (int iHeader = 0; iHeader<headers.length(); iHeader++)
+			{
+				String header = headers.getString(iHeader) + "\r\n";
+				out.write(header.getBytes());
+			}
+			out.write("\r\n".getBytes());
+			// output message body
+			out.write(messageContent.getBytes());
+			
+
+			messageStatus.put(Utils.JSON_ATTR_MAIL_MESSAGE_STATUS, Utils.STATUS_SUCCESS);
+		} catch (java.net.MalformedURLException e) {
+			// return status with error message
+			messageStatus.put(Utils.JSON_ATTR_MAIL_MESSAGE_STATUS, Utils.STATUS_FAILURE + " problem getting message content" + e.getMessage());
```

Comment:
Can other exception possible than `MalFormedURLException`? I feel it will be safe if you also have generic Exception in there